geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala

```diff
@@ -143,12 +145,53 @@ protected class QueryPlanFilterVisitor(sft: SimpleFeatureType) extends Duplicati
     }
   }
 
+  private def handleInArrayEquals(filter: PropertyIsEqualTo, extraData: AnyRef): Option[Filter] = {
+    var inArrayFunction: InArrayFunction = null
+    var hasTrue: Boolean = false
+
+    def inspectExpression(expression: Expression, extraData: AnyRef): Unit = {
+      expression match {
+        case function: InArrayFunction =>
+          inArrayFunction = function
+        case literal: Literal if java.lang.Boolean.TRUE.equals(literal.getValue) =>
+          hasTrue = true
+        case _ =>
+      }
+    }
+
+    inspectExpression(filter.getExpression1, extraData)
+    inspectExpression(filter.getExpression2, extraData)
+
+    if (inArrayFunction != null && hasTrue) {
+      val attribute = inArrayFunction.getParameters.get(0)
+      val array = inArrayFunction.getParameters.get(1)
+      import scala.collection.JavaConversions._
+      // The important part of this optimization is to handle ArrayLists generated by the
+      // GeoServer cross-layer querylayer extension:  E.g.
+      // https://github.com/geoserver/geoserver/blob/2.17.3/src/extension/querylayer/src/main/java/org/geoserver/filter/function/QueryFunction.java#L124-L151
```

Comment:
ah, thanks for this!

Reply:
Yeah, seemed like this was an important piece of the puzzle to comment on.  

Also, it'll invariably change later:).